go/teams/service_helper.go

```diff
@@ -448,65 +448,103 @@ type AddMembersRes struct {
 
 // AddMembers adds a bunch of people to a team. Assertions can contain usernames or social assertions.
 // Adds them all in a transaction so it's all or nothing.
-// On success, returns a list where len(res)=len(assertions) and in corresponding order.
-func AddMembers(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID, users []keybase1.UserRolePair) (res []AddMembersRes, err error) {
+// If the first transaction fails due to TeamContactSettingsBlock error, it
+// will remove restricted users returned by the error, and retry once.
+// On success, returns a list where len(added) + len(noAdded) = len(assertions) and in
+// corresponding order, with restricted users having an empty AddMembersRes.
+func AddMembers(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID, users []keybase1.UserRolePair) (added []AddMembersRes, notAdded []keybase1.User, err error) {
+	mctx := libkb.NewMetaContext(ctx, g)
 	tracer := g.CTimeTracer(ctx, "team.AddMembers", true)
 	defer tracer.Finish()
 
-	err = RetryIfPossible(ctx, g, func(ctx context.Context, _ int) error {
-		res = make([]AddMembersRes, len(users))
-		team, err := GetForTeamManagementByTeamID(ctx, g, teamID, true /*needAdmin*/)
-		if err != nil {
-			return err
-		}
+	addNonRestrictedMembersFunc := func(restrictedUsers map[keybase1.UID]bool) func(ctx context.Context, _ int) error {
+		return func(ctx context.Context, _ int) error {
```

Comment:
it looks like it can be simplified and `addNonRestrictedMembersFunc` can be just the function that's passed to `addNonRestrictedMembersFunc`, and it could use `restrictedUsers` by closing on a variable from outer `AddMembers` if it's declared and created early.

Reply:
your call, though, both seem fine. or maybe I'm missing something here